File: go/day2.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func checkLine(numLine []int) int {
	decrasing := numLine[0] > numLine[1]
	for i := range len(numLine) - 1 {
		if decrasing && numLine[i] < numLine[i+1] {
			return i
		}
		if !decrasing && numLine[i] > numLine[i+1] {
			return i
		}

		absDiff := math.Abs(float64(numLine[i] - numLine[i+1]))
		if absDiff > 3 || absDiff < 1 {
			return i
		}
	}
	return -1
}

func Day2a() {
	// read lines
	lines, err := ReadLines("../inputs/day2.txt")
	if err != nil {
		panic("Could not read lines")
	}

	// convert all numbers to integers
	numLines := make([][]int, 0)
	for _, line := range lines {
		nums := []int{}
		for _, numStr := range strings.Split(line, " ") {
			num, err := strconv.Atoi(numStr)
			if err != nil {
				panic("could not conver number")
			}
			nums = append(nums, num)
		}
		numLines = append(numLines, nums)
	}

	// Check
	count := 0
	for _, numLine := range numLines {
		mistake := checkLine(numLine)
		if mistake == -1 {
			count += 1
		}
	}

	fmt.Printf("%v\n", count)
}

func Day2b() {
	// read lines
	lines, err := ReadLines("../inputs/day2.txt")
	if err != nil {
		panic("Could not read lines")
	}

	// convert all numbers to integers
	numLines := make([][]int, 0)
	for _, line := range lines {
		nums := []int{}
		for _, numStr := range strings.Split(line, " ") {
			num, err := strconv.Atoi(numStr)
			if err != nil {
				panic("could not conver number")
			}
			nums = append(nums, num)
		}
		numLines = append(numLines, nums)
	}

	// Check
	count := 0
	for _, numLine := range numLines {
		mistake := checkLine(numLine)
		if mistake == -1 {
			count += 1
			continue
		}

		for _, removeInd := range []int{mistake - 1, mistake, mistake + 1} {
			if removeInd < 0 {
				continue
			}
			newLine := make([]int, len(numLine)-1)
			copy(newLine[:removeInd], numLine[:removeInd])
			copy(newLine[removeInd:], numLine[removeInd+1:])

			if checkLine(newLine) == -1 {
				count += 1
				break
			}
		}
	}

	fmt.Printf("%v\n", count)
}

File: python/day2.py
def check_line(line: list[int]) -> int:
    decreasing = line[0] > line[1]
    for i in range(len(line) - 1):
        if decreasing and line[i] < line[i + 1]:
            return i
        if not decreasing and line[i] > line[i + 1]:
            return i

        abs_diff = abs(line[i] - line[i + 1])
        if abs_diff < 1 or abs_diff > 3:
            return i

    return -1


def day2a() -> None:
    num_lines = []

    with open("../inputs/day2.txt", "r") as file:
        for line in file:
            parts = line.strip().split()
            num_lines.append([int(x) for x in parts])

    result = 0
    for line in num_lines:
        if check_line(line) == -1:
            result += 1

    print(result)

def day2b() -> None:
    num_lines = []

    with open("../inputs/day2.txt", "r") as file:
        for line in file:
            parts = line.strip().split()
            num_lines.append([int(x) for x in parts])

    result = 0
    for line in num_lines:
        mistake = check_line(line)
        if mistake == -1:
            result += 1
            continue

        for rem_ind in [mistake-1,mistake, mistake+1]:
            if rem_ind < 0:
                continue

            if check_line(line[:rem_ind] + line[rem_ind+1:]) == -1:
                result += 1
                break

    print(result)

day2b()
